### PR TITLE
Go with Redis integration

### PR DESCRIPTION
**Pull request by Cortey**
**Description**
The first iteration of Go with Redis integration implemented.

**Changes**

- Redis connection
- Dockerfile / Docker-Compose
- REST api
- .env support
- Database communication

Related Issue: #5

### CODE REVIEW

File: Backend/main.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/go-redis/redis/v8"
+	"github.com/gorilla/mux"
+)
+
+var (
```
Comment (akgalwas) [inline Backend/main.go:14]: I would avoid package level variables as they have global scope.

File: Backend/main.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+		return nil
+	}
+
+	return keys
+}
+
+func dbPostHandler(w http.ResponseWriter, r *http.Request) {
+	j, err := json.Marshal(Image{URL: "url", GCP: "gcp"}) //key: url | value: {url:"...", GCP:"..."}
+	if err != nil {
+		fmt.Println(err)
```
Comment (akgalwas) [inline Backend/main.go:67]: We should return InternalServer HTTP error here.

File: Backend/main.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	"github.com/gorilla/mux"
+)
+
+var (
+	connection = connectToRedis()
+	ctx        = context.Background()
+)
+
+type Image struct {
+	URL string `json:"url"`
```
Comment (mjakobczyk) [inline Backend/main.go:20]: If you expose variables in exposed struct you could provide a docs for it as well.

File: Backend/main.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	GCP string `json:"GCP"`
+}
+
+func connectToRedis() *redis.Client {
+
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     os.Getenv("REDIS_URL"),
```
Comment (mjakobczyk) [inline Backend/main.go:27]: What if there was no value provided for this env?

File: Backend/main.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+	return rdb
+}
+func insertToDB(key string, value string) error {
+	_, err := connection.Set(ctx, key, value, 0).Result()
+	if err != nil {
+		fmt.Println(err)
+	}
+	return err
+}
```
Comment (mjakobczyk) [inline Backend/main.go:44]: Missing empty line dividing function's definitions.

File: Backend/main.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+func getFromDB(key string) (interface{}, error) {
+	//val, err := connection.Do(ctx, "GET", key).Result()
+	val, err := connection.Get(ctx, key).Result()
+	if err != nil {
+		fmt.Println(err)
+	}
+	return val, err //fmt.Sprintf("%s", val)
```
Comment (mjakobczyk) [inline Backend/main.go:51]: Is this comment needed?

File: Backend/main.go [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func allKeys() []string {
```
Comment (mjakobczyk) [inline Backend/main.go:54]: What `allKeys` do you mean? Can you make the name of this function more precisely?

File: Backend/main.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		fmt.Println(err)
+		return nil
+	}
+
+	return keys
+}
+
+func dbPostHandler(w http.ResponseWriter, r *http.Request) {
+	j, err := json.Marshal(Image{URL: "url", GCP: "gcp"}) //key: url | value: {url:"...", GCP:"..."}
```
Comment (mjakobczyk) [inline Backend/main.go:65]: Is this comment needed?

File: Backend/main.go (hunk 2 above)
Comment (mjakobczyk) [inline Backend/main.go:67]: Please use logging instead of printing errors to standard output.

File: Backend/main.go [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+	if err != nil {
+		fmt.Println(err)
+	}
+
+	params := mux.Vars(r)
+	insertToDB(params["id"], string(j))
+
```
Comment (mjakobczyk) [inline Backend/main.go:72]: Unnecessary empty line.

File: Backend/main.go (hunk 8 above)
Comment (mjakobczyk) [inline Backend/main.go:65]: What does this variable `j` hold? Could you make its name more meaningful?

File: Backend/main.go [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func dbGetHandler(w http.ResponseWriter, r *http.Request) {
+	params := mux.Vars(r)
+	var x Image
+
+	if params["id"] == "999" {
```
Comment (mjakobczyk) [inline Backend/main.go:79]: What does this hardcoded `999` value mean? Could you make it a constant with a documentation?

File: Backend/main.go [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+		keys := allKeys()
+		for _, key := range keys {
+			fromDB, err := getFromDB(key)
+			if err != nil {
+				fmt.Println(err)
```
Comment (mjakobczyk) [inline Backend/main.go:84]: Please use logging instead of printing errors to standard output.

File: Backend/main.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+			} else {
+				fmt.Fprintf(w, "json = %s\n", fromDB)
+			}
+		}
+	}
+
+	fromDB, err := getFromDB(params["id"])
+	if err != nil {
+		fmt.Println(err)
```
Comment (mjakobczyk) [inline Backend/main.go:93]: Please use logging instead of printing errors to standard output.

File: Backend/main.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	err = json.Unmarshal([]byte(fromDB.(string)), &x)
+	if err != nil {
+		fmt.Println(err)
```
Comment (mjakobczyk) [inline Backend/main.go:98]: Please use logging instead of printing errors to standard output.

File: Backend/main.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+
+	fmt.Fprintf(w, "json = %s\n", fromDB)
+	fmt.Fprintf(w, "json URL = %s,  json GCP = %s\n", x.URL, x.GCP)
+
+}
+
+func main() {
+
+	if connection != nil {
+
```
Comment (mjakobczyk) [inline Backend/main.go:109]: Unnecessary empty line.

File: Backend/main.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+		router := mux.NewRouter()
+
+		router.HandleFunc("/{id}", dbGetHandler).Methods("GET")
+		router.HandleFunc("/{id}", dbPostHandler).Methods("POST")
+
+		fmt.Printf("Starting server at port 8081\n")
```
Comment (mjakobczyk) [inline Backend/main.go:115]: Please use logging instead of printing info to standard output.

File: Backend/main.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+		http.ListenAndServe(":8081", router)
```
Comment (mjakobczyk) [inline Backend/main.go:116]: Can this port be configurable?

File: Backend/makefile [hunk 17]
```diff
@@ -0,0 +1,4 @@
+build:
+	docker build -t image .
+push:
+
```
Comment (mjakobczyk) [inline Backend/makefile:4]: Is this target needed while it does not contain any instructions?

File: Backend/makefile [hunk 18]
```diff
@@ -0,0 +1,4 @@
+build:
+	docker build -t image .
```
Comment (mjakobczyk) [inline Backend/makefile:2]: `image` is too generic IMHO. You could name it more verbosely like: `redisClient`.

**Review by mjakobczyk — CHANGES_REQUESTED** [review]

**Comment by mjakobczyk** [comment]
Could you provide some unit tests for your code?

File: backend/database_test.go [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+import (
+	"context"
+	"errors"
+	"github.com/go-redis/redismock/v8"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestConnect(t *testing.T) {
+	t.Run("should check that connection is already established and return error", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:12]: You could name it this way:
`should return error when connection is already established`

File: backend/database_test.go [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+		client, _ := redismock.NewClientMock()
+		d := Database{address: "localhost:8081", password: "", connection: client, ctx: context.Background()}
+
+		//when
+		err := d.Connect()
+
+		//then
+		assert.Equal(t, nil, err)
+	})
+	t.Run("should return error when trying to Ping database", func(t *testing.T) { //todo:test refactor
```
Comment (mjakobczyk) [inline backend/database_test.go:23]: `Ping` does not have to be upper-cased

Comment (mjakobczyk) [inline backend/database_test.go:23]: Did you complete this `todo`?

File: backend/database_test.go [hunk 21]
```diff
@@ -0,0 +1,266 @@
...
+		err := d.Connect()
+
+		//then
+		assert.Error(t, err)
+		assert.NotEqual(t, nil, d.connection)
+	})
+}
+
+func TestInsertToDB(t *testing.T) {
+	t.Run("should return error when connection is not initialized", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:38]: You used `established` before, so it can be used here in place of `initialised` for consistency

File: backend/database_test.go [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+		database := Database{connection: client}
+		key, value := "key", "value"
+		clientMock.ExpectSet(key, value, 0).SetVal(value)
+
+		//when
+		err := database.InsertToDB(key, value)
+
+		//then
+		assert.NoError(t, err)
+		//clientMock.ExpectGet(key).SetVal(value)
```
Comment (mjakobczyk) [inline backend/database_test.go:87]: Can you remove not used comments?

File: backend/database_test.go [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+		//when
+		err := database.InsertToDB(key, value)
+
+		//then
+		assert.Error(t, err)
+		if err := clientMock.ExpectationsWereMet(); err != nil {
+			t.Error(err)
+		}
+	})
+	t.Run("should insert data correctly using the InsertToDB function - returns no error", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:72]: It can be named this way:
`should insert data correctly with no errors`

File: backend/database_test.go [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+		//assert.NoError(t, err)
+		//assert.Equal(t, value, val)
+		if err := clientMock.ExpectationsWereMet(); err != nil {
+			t.Error(err)
+		}
+	})
+}
+
+func TestGetFromDB(t *testing.T) {
+	t.Run("should return error when connection is not initialized", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:98]: You use `established` or `initialised` - try to stick to one naming for consistency

File: backend/database_test.go [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+		val, err := database.GetFromDB(key)
+
+		//then
+		assert.Equal(t, "", val)
+		assert.Equal(t, "GETFROMDB:for key " + key + " value is empty", err.Error())
+		if err := clientMock.ExpectationsWereMet(); err != nil {
+			t.Error(err)
+		}
+	})
+	t.Run("should acquire data correctly using GetFromDB function", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:171]: Instead of `acquire` simply use `get` word as it is the name of the function

File: backend/database_test.go [hunk 26]
```diff
@@ -0,0 +1,266 @@
...
+			t.Error(err)
+		}
+	})
+	t.Run("should return error when key doesn't exists in database", func(t *testing.T) {
+		//given
+		client, clientMock := redismock.NewClientMock()
+		clientMock.ClearExpect()
+		clientMock.MatchExpectationsInOrder(true)
+		database := Database{connection: client}
+		key := "key"
```
Comment (mjakobczyk) [inline backend/database_test.go:120]: This variable can be shared on test-scope level, not only single run

File: backend/database_test.go [hunk 27]
```diff
@@ -0,0 +1,266 @@
...
+
+		//then
+		assert.Error(t, err)
+		var n []string = nil
+		assert.Equal(t, n, val)
+		if err := clientMock.ExpectationsWereMet(); err != nil {
+			t.Error(err)
+		}
+	})
+	t.Run("should acquire all keys using the GetAllKeys function", func(t *testing.T) {
```
Comment (mjakobczyk) [inline backend/database_test.go:228]: As above - function is called `get` so why not using this word instead of `acquire` in test description?

Comment (mjakobczyk) [inline backend/database_test.go:228]: Do not state verbosely `using the GetAllKeys function` because the function name can change - it is only implementation, test descriptions should be high-level and generic

File: backend/docker-compose.yml [hunk 28]
```diff
@@ -0,0 +1,22 @@
...
+        image: "redis:alpine"    
+        command: redis-server #--requirepass ${REDIS_PASSWORD}
+        ports:
+         - "6379:6379"
+        environment:
+         - REDIS_REPLICATION_MODE=master
+        container_name: redis
+    go:
+        build: .
+        image: image # docker build -t image .
```
Comment (mjakobczyk) [inline backend/docker-compose.yml:13]: Is this comment required?

File: backend/docker-compose.yml [hunk 29]
```diff
@@ -0,0 +1,22 @@
+version: '3.1'
+services: 
+    redis:
+        image: "redis:alpine"    
+        command: redis-server #--requirepass ${REDIS_PASSWORD}
```
Comment (mjakobczyk) [inline backend/docker-compose.yml:5]: Is this comment required?

File: backend/main.go [hunk 30]
```diff
@@ -0,0 +1,54 @@
...
+	"net/http"
+	"os"
+)
+
+// initAPIHandler initializes a handler for the API.
+func initAPIHandler() (Handler, error) {
+	address := os.Getenv("REDIS_URL")
+	if address == "" {
+		log.Info("Failed to read REDIS_URL from .env file, using a default value instead")
+		address = "redis:6379" //todo Czy pozwalać programowi kontynuować? Czy lepiej przerwać?
```
Comment (mjakobczyk) [inline backend/main.go:15]: Did you complete this TODO?

File: backend/main.go [hunk 31]
```diff
@@ -0,0 +1,54 @@
...
+
+	router.HandleFunc("/v1/images", handler.DBGetAllHandler).Methods("GET")
+	router.HandleFunc("/v1/images/{id}", handler.DBGetHandler).Methods("GET")
+	router.HandleFunc("/v1/images/{id}", handler.DBPostHandler).Methods("POST")
+
+	port := os.Getenv("PORT")
+	if port == "" {
+		port = "8081"
+		log.Info("Failed to read PORT from .env file, using a default value instead")
+		//todo return?? To samo co z REDIS_URL
```
Comment (mjakobczyk) [inline backend/main.go:44]: Did you complete this TODO?

File: backend/model.go [hunk 32]
```diff
@@ -0,0 +1,109 @@
+package main
+
+// Image defines a struct to use for the received image
+type Image struct {
+	URL string `json:"url"`
+	GCP string `json:"gcp"`
+	IMG string `json:"img"`
+}
+
+// func toBase64(b []byte) string {
```
Comment (mjakobczyk) [inline backend/model.go:10]: Do you need all these commented code?

File: backend/tests/database-connection_test.go [hunk 33]
```diff
@@ -0,0 +1,5 @@
+package tests
+
+type User struct {
```
Comment (mjakobczyk) [inline backend/tests/database-connection_test.go:3]: This type lacks docs

File: backend/tests/database-connection_test.go [hunk 34]
```diff
@@ -0,0 +1,5 @@
...
+	Username string `j`
```
Comment (mjakobczyk) [inline backend/tests/database-connection_test.go:4]: This variable lacks docs

File: backend/tests/endpoint_test.go [hunk 35]
```diff
@@ -0,0 +1,12 @@
+package tests
+
+//func TestGetEntries(t *testing.T){
```
Comment (mjakobczyk) [inline backend/tests/endpoint_test.go:3]: Is this file needed? You provided no tests at all

**Review by mjakobczyk — CHANGES_REQUESTED** [review]

File: frontend/public/index.html [hunk 36]
```diff
@@ -1,15 +0,0 @@
...
-    <meta name="viewport" content="width=device-width, initial-scale=1" />
-    <meta name="theme-color" content="#000000" />
-    <meta name="description" content="Kyma Showcase" />
-    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
-    <title>Kyma showcase</title>
-  </head>
-  <body>
-    <div id="root"></div>
-  </body>
-</html>
```
Comment (mwisnia97) [inline frontend/public/index.html:15]: I think this file is needed.

**Comment by ksputo** [comment]
Please birng the frontend back

**Comment by ksputo** [comment]
Let's discuss the packages - it is not a good practice to place everything in 'package main'

**Comment by Cortey** [comment]
> Let's discuss the packages - it is not a good practice to place everything in 'package main'

Thank you for the tip, we will do it in the next PR.

File: RESTAPI_contract.md [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+`GET v1/images/{id}/description`\
+The endpoint returns the following JSON
+`[{ id: "#",
+title: "#",
+content: base64,
+description: {JSON from GCP},
+status: "#"
+}]`\
+The endpoint returns the following HTTP codes:
+- 200 (if processing is done)
```
Comment (ksputo) [inline RESTAPI_contract.md:34]: Consider adding `404 Not Found` when trying to get non-existing resource. See https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.4

File: backend/api.go [hunk 38]
```diff
@@ -0,0 +1,147 @@
+package main
```
Comment (ksputo) [inline backend/api.go:1]: I'd like you to work on domain naming - firstly, think of package name that fits the domain use case of this code. Then, create a new directory, for example `internal/database` and move this file there - with package name changed to `database`. Also, the filename could be more informative as `api` is too general.

See:
- https://golang.org/doc/effective_go#package-names

File: backend/database.go [hunk 39]
```diff
@@ -0,0 +1,86 @@
...
+	})
+
+	if err := d.connection.Ping(d.ctx).Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// InsertToDB adds a database entry using provided key(String) and value(String).
+func (d Database) InsertToDB(key string, value string) error {
```
Comment (ksputo) [inline backend/database.go:48]: You're already calling this method on type `Database` - consider renaming methods to avoid repetitions like:

db := Database{}
**db**.InsertTo**DB**


imo just `Insert()` would be enough, cause you always call it on `Database` type - think this way - where else could you insert data than `ToDB`?

**Review by mjakobczyk — APPROVED** [review]
LGTM. Make sure to do mentioned minor changes in the follow-up as we talked. 👍

**Review by mwisnia97 — APPROVED** [review]
LGTM

File: backend/api.go (hunk 38 above)
Comment (mvshao) [reply backend/api.go:1]: Added to the next sprint.
